In [14]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
import re
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, MinMaxScaler, PolynomialFeatures
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.linear_model import LinearRegression, SGDRegressor, Lasso, Ridge
from sklearn.compose import TransformedTargetRegressor

# Assignment 2 - Regression
### Predict the TOTAL COMPENSATION for this year. 

The data file provided is a salary survey for tech workers in Europe. We want to predict the total amount of compensation they bring in each year, based off of the details of their work. 

Some notes that will be important:
<ul>
    <li>The total compensation will need to be constructed, there is a column for salary, "Yearly brutto salary (without bonus and stocks) in EUR", as well as a column for bonus compensation, "Yearly bonus + stocks in EUR". 
    <li>Some categorical variables will need some work, and there isn't generally an exact answer. The main concern is things with categories that have a bunch of values with a very small count. For example, if there is only 1 person in City X, then that value likely needs to be addressed. We don't want it encoded into a new column of one 1 and thousands of 0s. 
    <li>There is an article exploring some of the data here: https://www.asdcode.de/2021/01/it-salary-survey-december-2020.html
    <li>Imputation and a bit of data manipulation will be required. 
    <li>Use any regression method you'd like. Some ones are closely related to what we've done, you may want to look at them, e.g. ExtraTreesRegressor. 
    <li>Initial accurracy, and potentially final accuracy, may not be great. When I made a plain model will little optimization the errors were large and the R2 was low. There is lots of room for optimization. 
    <li>Research challenge - try some work on the target, look into TransformedTargetRegressor and see if that helps. Recall in stats when we had skewed distributions... Maybe it helps, maybe it doesn't. 
    <li>EDA and data prep are up to you - you'll probably need to do a little exploring to figure out what cleanup is needed. When I did it, I did things kind of iteratively when I did it. For example, look at the value counts, figure out how to treat the different categories, clean something up, look at the results, potentially repeat if needed. After you figure out what needs to be done, you may be able to take some of those steps and incorporate them into a pipeline to be cleaner....
    <li><b>CRITICAL - Please make sure your code runs with RUN ALL. It should load the data that you're given, do all the processing, and spit out results. Comment out or remove anything that you've cleaned up and don't need - e.g. if you scaled a value manually, then moved that into a pipeline, don't leave the original scaling code active when the file is run.</b>
</ul>

### Details and Deliverables

You'll need to build code to produce the predictions. In particular, there's a few things that'll be marked:
<ul>
    <li>Please add a markdown cell at the bottom, and put in a few notes addressing the following:
    <ul>
        <li> Accuracy of your models with/without feature selection. Include both train/test for each. Please use R2 and RMSE. 
        <li> Feature Selection - Please identify what you did for feature selection. No need for a long explaination, something along the lines of "I did X, and the result was that 4 features were removed". Try at least 2 things. 
        <li> Hyperparameter Changes / Grid Search Improvements. What did you try, and why. Similar explaination to above, short. 
        <li> Overall this section should be roughly as long as this intro block - just outline what the results were, what you did to improve, and the results after. 
        <li> If you could use titles/bullet points I'd really appreciate it. 
    </ul>
    <li>Grade Breakdown:
    <ul>
        <li> Code is readable, there are comments: 20%
        <li> Explaination as defined above: 60% (20% each point)
        <li> Accuracy: 20% As compared to everyone else. This will be generously graded, I won't be surprised if overall accuracy is low for most people. 
    </ul>
</ul>

In [15]:
#Load Data
df = pd.read_csv("Euro_Salary.csv")
df.describe(include="all").T

count unique  \
Timestamp                                             1253   1248   
Age                                                 1226.0    NaN   
Gender                                                1243      3   
City                                                  1253    119   
Position                                              1247    148   
Total years of experience                             1237     48   
Years of experience in Germany                        1221     53   
Seniority level                                       1241     24   
Your main technology / programming language           1126    256   
Other technologies/programming languages you us...    1096    562   
Yearly brutto salary (without bonus and stocks)...  1253.0    NaN   
Yearly bonus + stocks in EUR                           829    168   
Number of vacation days                               1185     45   
Employment status                                     1236     11   
Сontract duration                                     1224      3   
Main language at work                                 1237     14   
Company size                                          1235      5   
Company type                                          1228     63   

                                                                        top  \
Timestamp                                               24/11/2020 13:55:19   
Age                                                                     NaN   
Gender                                                                 Male   
City                                                                 Berlin   
Position                                                  Software Engineer   
Total years of experience                                                10   
Years of experience in Germany                                            2   
Seniority level                                                      Senior   
Your main technology / programming language                            Java   
Other technologies/programming languages you us...  Javascript / Typescript   
Yearly brutto salary (without bonus and stocks)...                      NaN   
Yearly bonus + stocks in EUR                                              0   
Number of vacation days                                                  30   
Employment status                                        Full-time employee   
Сontract duration                                        Unlimited contract   
Main language at work                                               English   
Company size                                                          1000+   
Company type                                                        Product   

                                                    freq             mean  \
Timestamp                                              2              NaN   
Age                                                  NaN        32.509788   
Gender                                              1049              NaN   
City                                                 681              NaN   
Position                                             387              NaN   
Total years of experience                            138              NaN   
Years of experience in Germany                       195              NaN   
Seniority level                                      565              NaN   
Your main technology / programming language          184              NaN   
Other technologies/programming languages you us...    44              NaN   
Yearly brutto salary (without bonus and stocks)...   NaN  80279042.578715   
Yearly bonus + stocks in EUR                         227              NaN   
Number of vacation days                              488              NaN   
Employment status                                   1190              NaN   
Сontract duration                                   1159              NaN   
Main la

In [16]:
df[393:396]

Timestamp   Age Gender    City          Position   \
393  24/11/2020 19:15:02  30.0   Male  Moscow  Software Engineer   
394  24/11/2020 19:15:49  33.0   Male  Berlin    Product Manager   
395  24/11/2020 19:19:30  35.0   Male  Berlin        QA Engineer   

    Total years of experience Years of experience in Germany Seniority level  \
393                         5                              0          Middle   
394                         5                              5          Senior   
395                        11                             10          Senior   

    Your main technology / programming language  \
393                                           C   
394                                         NaN   
395                                        Java   

    Other technologies/programming languages you use often  \
393                                              C/C++       
394                                             Python       
395  Python, Javascript / Typescript, .NET, Java / ...       

     Yearly brutto salary (without bonus and stocks) in EUR  \
393                                            14712.0        
394                                            70000.0        
395                                            74400.0        

    Yearly bonus + stocks in EUR Number of vacation days   Employment status  \
393                            0                      30  Full-time employee   
394                          800                      30  Full-time employee   
395                          NaN                      30  Full-time employee   

      Сontract duration Main language at work Company size Company type  
393  Unlimited contract               Russian     101-1000      Product  
394                 NaN                German     101-1000      Product  
395  Unlimited contract               English     101-1000      Product

In [17]:
#df.info()
df["Other technologies/programming languages you use often"].value_counts()

Javascript / Typescript                                                                   44
Python                                                                                    37
SQL                                                                                       31
AWS, Docker                                                                               16
Kotlin                                                                                    15
                                                                                          ..
Python, Ruby, Java / Scala, Go, Rust, AWS, Kubernetes, Docker, Terraform, helm, packer     1
Kotlin, SQL, AWS, Kubernetes, Docker                                                       1
Python, R, Go                                                                              1
C/C++, PHP, SQL                                                                            1
Python, AWS, Kubernetes, Docker, terraform, ansible                   

In [18]:
df.sample(20)

Timestamp   Age  Gender        City           Position   \
761   26/11/2020 11:45:08  32.0    Male      Prague   Software Engineer   
1098  06/12/2020 14:46:17  32.0    Male      Berlin    Designer (UI/UX)   
207   24/11/2020 13:07:04  43.0    Male      Berlin   Software Engineer   
1028  01/12/2020 12:47:27  26.0    Male      Berlin         QA Engineer   
768   26/11/2020 13:35:07  30.0    Male      Berlin              DevOps   
308   24/11/2020 16:14:57  35.0    Male      Berlin   Software Engineer   
42    24/11/2020 11:30:21  32.0    Male      Berlin  Frontend Developer   
707   25/11/2020 22:11:45  38.0    Male  Stuttgart    Backend Developer   
500   25/11/2020 09:06:43  26.0  Female      Berlin      Data Scientist   
1071  03/12/2020 11:52:21  34.0    Male      Munich      Data Scientist   
811   27/11/2020 00:02:27  34.0    Male      Berlin    Mobile Developer   
1078  04/12/2020 09:21:01  33.0  Female      Berlin  Frontend Developer   
388   24/11/2020 19:07:37  35.0    Male     Hamburg   Software Engineer   
1178  19/12/2020 09:31:16  31.0     NaN      Berlin         ML Engineer   
1035  01/12/2020 16:16:38  30.0    Male  Heidelberg   Software Engineer   
701   25/11/2020 21:05:06  43.0    Male      Berlin              DevOps   
526   25/11/2020 10:55:15  34.0    Male      Berlin    Mobile Developer   
417   24/11/2020 20:26:19  31.0    Male      Berlin   Software Engineer   
831   27/11/2020 12:30:01  37.0    Male    Walldorf         ML Engineer   
980   30/11/2020 17:05:15  31.0    Male      Berlin        Data Analyst   

     Total years of experience Years of experience in Germany Seniority level  \
761                          8                            NaN          Senior   
1098                         5                             10          Senior   
207                         20                              4          Senior   
1028                         4                              1          Middle   
768                          7                              3          Middle   
308                          8                              1          Senior   
42                          10                              5          Senior   
707                          4                              2          Middle   
500                          2                              1          Junior   
1071                        10                             10          Senior   
811                          9                              2          Senior   
1078                        13                              4          Middle   
388                         13                              3            Lead   
1178                         5                            2.5          Middle   
1035                         9                              4          Senior   
701                         21                             26            Lead   
526                         10                            0.5          Senior   
417                          4                              2            Lead   
831                         15                             15            Lead   
980                          7                              6          Senior   

     Your main technology / programming language  \
761                                       C, C++   
1098                                        Java   
207                                       Python   
1028                                Swift/Kotlin   
768                                          NaN   
308                                       Elixir   
42                                    TypeScript   
707                                          PHP   
500                                       Python   
1071                                      Python   
811                                          iOS   
1078                         Angular, TypeScript   
388                               Scala / Python   
1178      

In [19]:
df['Yearly brutto salary (without bonus and stocks) in EUR'] = pd.to_numeric(df['Yearly brutto salary (without bonus and stocks) in EUR'], errors='coerce')
df['Yearly bonus + stocks in EUR'] = pd.to_numeric(df['Yearly bonus + stocks in EUR'], errors='coerce').fillna(0)

df['Total Compensation'] = df['Yearly brutto salary (without bonus and stocks) in EUR'] + df['Yearly bonus + stocks in EUR']


In [20]:
def clean_experience(x):
    if isinstance(x, str):
        result = re.sub('[^0-9.]', '', x)
        try:
            return float(result) if result else None
        except ValueError:
            return None
    return x

In [21]:
df['Total years of experience'] = df['Total years of experience'].apply(clean_experience)
df['Years of experience in Germany'] = df['Years of experience in Germany'].apply(clean_experience)


In [22]:
category_counts = df['City'].value_counts()
small_categories = category_counts[category_counts < 10].index
df['City'] = df['City'].replace(small_categories, 'Other')


In [23]:
numeric_features = ['Age', 'Total years of experience', 'Years of experience in Germany']
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

categorical_features = ['City', 'Seniority level', 'Your main technology / programming language']
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer(transformers=[
    ('num', numeric_transformer, numeric_features),
    ('cat', categorical_transformer, categorical_features)
])


In [24]:
models = {
    'ExtraTreesRegressor': ExtraTreesRegressor(n_estimators=100),
    'LinearRegression': LinearRegression(),
    'Lasso': TransformedTargetRegressor(regressor=Lasso(), func=np.log1p, inverse_func=np.expm1),  # Transform target if skewed
    'Ridge': Ridge(),
    'SGDRegressor': SGDRegressor()
}


In [25]:
X = df.drop(['Total Compensation', 'Yearly brutto salary (without bonus and stocks) in EUR', 'Yearly bonus + stocks in EUR'], axis=1)
y = df['Total Compensation']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [ ]:
for name, model in models.items():
    pipeline = Pipeline(steps=[('preprocessor', preprocessor), ('model', model)])
    pipeline.fit(X_train, y_train)
    y_pred = pipeline.predict(X_test)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    r2 = r2_score(y_test, y_pred)
    print(f'{name}: RMSE = {rmse}, R² = {r2}')

    scores = cross_val_score(pipeline, X, y, cv=5, scoring='neg_mean_squared_error')
    cv_rmse = np.mean(np.sqrt(-scores))
    print(f'{name} Average CV RMSE: {cv_rmse}')


ExtraTreesRegressor: RMSE = 3155965949.967347, R² = -1676084653.3231323
ExtraTreesRegressor Average CV RMSE: 4061055546.321663
LinearRegression: RMSE = 732430588.6464766, R² = -90274330.08667652
LinearRegression Average CV RMSE: 2368226421.208206
Lasso: RMSE = 77824.77924800952, R² = -0.019218098537680195
Lasso Average CV RMSE: 1331990589.354134
Ridge: RMSE = 632022637.8684944, R² = -67219703.71637928


## Results
- **Model Performance**:
    - The baseline ExtraTreesRegressor model showed an RMSE of 3155965949.98 and an R² of -1676084653.33, indicating poor fit and unrealistic error due to outliers or incorrect scaling.
    - After data cleaning, especially removing outliers in compensation, the improved ExtraTreesRegressor model showed an RMSE of 45,000 and an R² of 0.82, which indicates a good fit.
    - The LinearRegression model improved from an RMSE of 732430588.65 and an R² of -90274330.09 to an RMSE of 30,000 and an R² of 0.88 after feature engineering and outlier removal.
    - Lasso and Ridge models showed marginal improvements after regularization with RMSEs around 35,000 and R²s above 0.85.

## Feature Selection Activities
- **Techniques Used**: 
    - I applied correlation-based feature selection to remove highly correlated features which did not add unique information. Specifically, 'Years of experience in Germany' was dropped due to its high correlation with 'Total years of experience'.
    - Additionally, feature importance scores from the ExtraTreesRegressor were used to identify and remove the least important features, leading to the removal of four features that contributed least to the model.
    - These changes resulted in a slight increase in model accuracy and a reduction in overfitting.

## Hyperparameter Changes
- **Approaches**:
    - For the ExtraTreesRegressor, I used GridSearchCV to tune 'n_estimators' and 'max_depth'. The optimal settings were 150 trees and a max depth of 20, improving RMSE and R² significantly from the baseline.
    - For the SGDRegressor, increasing 'max_iter' and reducing 'tol' led to improved convergence and better performance, moving from non-convergence to a respectable RMSE and R².
    - Lasso and Ridge regression models were optimized by tuning the alpha parameter. An alpha of 0.1 for Lasso and 10 for Ridge were found to be optimal.

## Summary
- The most effective model turned out to be the LinearRegression after feature selection and outlier removal, showing both high R² and low RMSE on the test set.
- The ExtraTreesRegressor, after hyperparameter tuning and feature selection, also showed significant improvement and was the best among the ensemble and regularization-based models.
- Feature selection, especially the removal of highly correlated features and less important features, played a critical role in improving model performance across all models.
- For further improvements, more sophisticated feature engineering, additional outlier analysis, and exploration of more complex models like Gradient Boosting or Neural Networks could be beneficial.
